gamemode/modules/base/sv_gamemode_functions.lua

```diff
@@ -304,36 +304,36 @@ local vrad = GM.Config.voiceradius
 local dynv = GM.Config.dynamicvoice
 local deadv = GM.Config.deadvoice
 local voiceDistance = GM.Config.voiceDistance * GM.Config.voiceDistance
+local DrpCanHear = {}
 -- proxy function to take load from PlayerCanHearPlayersVoice, which is called a quadratic amount of times per tick,
 -- causing a lagfest when there are many players
 local function calcPlyCanHearPlayerVoice(listener)
-    if not IsValid(listener) then return end
-    listener.DrpCanHear = listener.DrpCanHear or {}
+    DrpCanHear[listener] = DrpCanHear[listener] or {}
     local shootPos = listener:GetShootPos()
     for _, talker in ipairs(player.GetAll()) do
         local talkerShootPos = talker:GetShootPos()
-        listener.DrpCanHear[talker] = not vrad or -- Voiceradius is off, everyone can hear everyone
+        DrpCanHear[listener][talker] = not vrad or -- Voiceradius is off, everyone can hear everyone
             (shootPos:DistToSqr(talkerShootPos) < voiceDistance and -- voiceradius is on and the two are within hearing distance
                 (not dynv or IsInRoom(shootPos, talkerShootPos, talker))) -- Dynamic voice is on and players are in the same room
     end
 end
 hook.Add("PlayerInitialSpawn", "DarkRPCanHearVoice", function(ply)
+    calcPlyCanHearPlayerVoice(ply)
     timer.Create(ply:UserID() .. "DarkRPCanHearPlayersVoice", 0.5, 0, fn.Curry(calcPlyCanHearPlayerVoice, 2)(ply))
 end)
+
 hook.Add("PlayerDisconnected", "DarkRPCanHearVoice", function(ply)
-    if not ply.DrpCanHear then return end
```

Comment:
`DrpCanHear[ply] = nil`?

Reply:
It cannot be nil now.
It simply cannot.

Reply:
That's not what I meant. I meant you should SET `DrpCanHear[ply] = nil`. You're setting the table for all other players, but not for the leaving player.

Say you have `ply1`, `ply2`, `ply3`. At some point the table looks like this:

```lua
{ [ply1] = {[ply2] = true, [ply3] = true}
, [ply2] = {[ply1] = true, [ply3] = true}
, [ply3] = {[ply1] = true, [ply2] = true}
}
```

When `ply1` leaves, you want `DrpCanHear[ply1]` to be nil, besides making `DrpCanHear[ply2][ply1]` and `DrpCanHear[ply3][ply1]` nil. Then you should have a table with two members, each holding a table of one value:

```lua
{ [ply2] = {[ply3] = true}
, [ply3] = {[ply2] = true}
}
```

I'd like you to run some tests to make sure this works correctly.

Also, are you 100% sure that the `PlayerInitialSpawn` is called before `player.GetAll()` can hold the player anywhere? I wouldn't be so sure. GMod has whacky stuff like that.

Reply:
PlayerInitialSpawn is called after PlayerAuthed. It should be guaranteed player.GetAll() will contain them.

Reply:
> PlayerInitialSpawn is called after PlayerAuthed. It should be guaranteed player.GetAll() will contain them.

`PlayerInitialSpawn` being called *after* things is even more worrying. I worry about this happening:

- Ply1 joins
- Ply2 disconnects
- `PlayerDisconnected` gets called for `ply2`
- `Ply1` is in `player.GetAll()`, but `DrpCanHear[ply1]` is `nil`, so you get a nil error
- `PlayerInitialSpawn` gets called for `ply1`, which sets `DrpCanHear[ply1]`

Don't prove to me `PlayerInitialSpawn` is called AFTER something. Prove to me that is definitely called BEFORE the `PlayerDisconnected` of another player.

Reply:
InitialSpawn is always called before Disconnected, I tested.
With a bot.

Reply:
This is a very timing sensitive thing. Something that may take a year before someone reports the error. How did you test this?

Reply:
Spawning the bot.
Next frame, kick them.
It's literally a frame away.

Reply:
I'm still having a bad stomach feeling about this. Years of people reporting the most insane errors made me paranoid. 

Let me as the @Kefta guru to be sure: I don't develop for gmod nearly enough to be sure about subtle foot guns like this. Do you think there's a risk of the described error?

Reply:
Even if there is.
It corrects it within 0.5 seconds.
There is no risk here.
Even if it errors, we can quickly make a change within seconds.

Reply:
Yeah, you're going to have a hard time explaining that to the server owner who doesn't have a clue on how bad the error is or what it even means. The risk is server owners being stressed out about errors that we are anticipating right here. When they see the error we've already lost, regardless of whether the error has technical consequences.

Also, such errors degrade the perceived quality of DarkRP. It shouldn't throw errors for issues not caused by the server owner.

Reply:
I am running it on 3 servers without any issues.
If it errors, I see it first.
An error which does nothing, does no harm, because it's easily fixed.

We just had an error caused by an update and it did no harm at all, because I reported it within a day.

Reply:
No, you're explicitly **not** the first to see it. This error would be extremely rare. It's the kind that gets reported YEARS after being introduced. Thinking that you will be the one to find it, and soon even, is extremely naive.

The check isn't even in a hot code path. It's in the PlayerDisconnected hook.

Reply:
I'm running it on 3 servers, simultaneously, pre launch.
I'm a developer which is frequently checking their console.

If the owner of the server doesn't fuck up PlayerDisconnected, there will be no error.
I've even tested further, with kicking on PlayerAuthed, which then resulted in being still valid in PlayerInitialSpawn and calling PlayerDisconnected properly.

The only edge case scenario which even exists is, if the player errors on PlayerDisconnected, because then, the timer wouldn't discontinue, when the player disconnects.

Additionally PlayerAuthed isn't called for bots, AT ALL, which is essencially.

HERE:
https://pastebin.com/sapQAwAw -- Tested with a real player AND a bot.
Both didn't error.

Code:
https://pastebin.com/r9YNvSUP